more work towards getting rid of client.Call

tbg opened the pull request:
replaces `client.Sender` with the artist formerly known as `client.BatchSender`.
Some usage of `client.Call` remains, mostly in `client` and in test code (to be tackled next).

@bdarnell feel free to reassign (you made the mistake of looking at the last one :smile:)

tbg committed:
use BatchSender in kv.DB

tbg committed:
use client.BatchSender everywhere

also took primary authorship of TxnCoordSender.

tbg committed:
unwrap in a few more places

tbg committed:
remove client.Sender

tschottdorf commented on client/rpc_sender.go:
```diff
@@ -74,16 +74,17 @@ func newRPCSender(server string, context *base.Context, retryOpts retry.Options,
 }
 
-// Send sends call to Cockroach via an RPC request. Errors which are retryable
+// SendBatch send a request to Cockroach via RPC. Errors which are retryable
```
not sure what happened here. Going to fix the comment.

tschottdorf replied:
done.

tbg committed:
client.{BatchSender->Sender}

and client.Sender#{SendBatch->Send}. Some "call"-based was removed
in the process but a little bit is left.

unexported provisional SendCallConverted in favor of SendCall (which does the
conversion and also auto-wraps in a batch).

tamird commented on client/client_test.go:
```diff
@@ -69,11 +69,16 @@ func (ss *notifyingSender) wait() {
+func (ss *notifyingSender) Send(ctx context.Context, ba proto.BatchRequest) (*proto.BatchResponse, *proto.Error) {
+	br, pErr := ss.wrapped.Send(ctx, ba)
+	if br != nil && br.Error != nil {
+		log.Warningf("culprit is %T", ss.wrapped)
```
just fmt.Sprintf this into the panic

tschottdorf replied:
done.

mrtracy commented on client/rpc_sender.go:
```diff
@@ -109,6 +111,9 @@ func (s *rpcSender) Send(_ context.Context, call proto.Call) {
-		call.Reply.Header().SetGoError(err)
+		return nil, proto.NewError(err)
 	}
+	pErr := br.Error
```
This pattern bothers me a bit (error on the BatchResponse only used to pass error up from Call), is this one of the things that will be changing?

tamird replied:
at this point, we're on the wire, so there's really no way around this so long as our protos have an error field.

tschottdorf replied:
yep, though we can make it invisible to everything else if we move it into a message that holds a batch request and an error. It's one of the things I'm going to consider when this stuff here is through.

tamird commented on client/db_internal_test.go:
```diff
@@ -29,11 +29,12 @@ import (
 		count++
-		if call.Args.Header().CmdID.WallTime == 0 {
+		if ba.CmdID.WallTime == 0 {
 			t.Errorf("expected client command ID to be initialized")
```
perhaps take this opportunity to return an error

tschottdorf replied:
done.

tamird commented on client/provisional.go:
```diff
@@ -45,13 +60,14 @@ func SendCallConverted(sender BatchSender, ctx context.Context, call proto.Call)
 		gogoproto.Merge(call.Reply, reply)
 	}
 	if call.Reply.Header().GoError() != nil {
+		log.Warningf("culprit is %T", sender)
```
ditto about sprintf

tschottdorf replied:
done.

tamird commented on client/rpc_sender.go:
```diff
@@ -74,16 +74,17 @@ func newRPCSender(server string, context *base.Context, retryOpts retry.Options,
+// and been executed successfully. We retry here to eventually get through with
+// the same client command ID and be given the cached response.
+func (s *rpcSender) Send(ctx context.Context, ba proto.BatchRequest) (*proto.BatchResponse, *proto.Error) {
+	method := fmt.Sprintf("Server.%s", proto.Batch)
```
this can live outside this method, yeah?

tschottdorf replied:
done.

tamird commented on client/rpc_sender.go:
```diff
@@ -93,7 +94,8 @@ func (s *rpcSender) Send(_ context.Context, call proto.Call) {
 		}
 
-		if err = s.client.Call(method, call.Args, call.Reply); err != nil {
+		br = &proto.BatchResponse{}
```
`var br BatchResponse` above instead of allocating this on every retry

tschottdorf replied:
done (guess we can trust the client to not touch the response if it fails?)

tamird replied:
or you can do a nil check on the error before returning

tschottdorf replied:
I'm just calling `.Reset()` in the error case.

tamird commented on client/txn.go:
```diff
@@ -44,21 +44,26 @@ var DefaultTxnRetryOptions = retry.Options{
+	br, pErr := ts.wrapped.Send(ctx, ba)
+	if br != nil && br.Error != nil {
+		log.Warningf("culprit is %T", ts.wrapped)
+		panic(proto.ErrorUnexpectedlySet)
```
sprintf

tschottdorf replied:
done.

tamird commented on client/txn_test.go:
```diff
@@ -101,12 +109,11 @@ func newTestSender(pre, post func(proto.Call)) SenderFunc {
-			WallTime: walltime,
-			Logical:  logical,
-		}
+		return proto.ZeroTimestamp.Add(walltime, logical)
```
sadly, there is no :pedantic: emoji

tschottdorf replied:
but there is: ![](https://avatars3.githubusercontent.com/u/1535036?v=3&s=20)

:dragon_face:

tamird commented on kv/local_test_cluster.go:
```diff
@@ -81,12 +81,16 @@ func (ltc *LocalTestCluster) Start(t util.Tester) {
-		ltc.localSender.Send(context.Background(), call)
-		return []gogoproto.Message{call.Reply}, call.Reply.Header().GoError()
+		if br.Error != nil {
+			panic(proto.ErrorUnexpectedlySet)
```
no culprit logging? perhaps proto should just provide a method that panics conveniently

tschottdorf replied:
I was planning to remove most of these eventually. I added them only because I ran into that during refactor and it was most convenient to track it down that way. I'm going to touch this up a bit.

tamird commented on storage/store_test.go:
```diff
@@ -82,46 +82,37 @@ type responseAdder interface {
+	}
+	ba.Replica = *replica
+	reply, pErr := db.store.ExecuteCmd(ctx, &ba)
+	br, _ := reply.(*proto.BatchResponse)
```
isn't it always a batchreponse?

tschottdorf replied:
not if it's `nil` (i.e. on error).

tamird replied:
Ah, right.
On Sep 22, 2015 5:45 PM, "Tobias Schottdorf" <notifications@github.com>
wrote:

> In storage/store_test.go
> <https://github.com/cockroachdb/cockroach/pull/2622#discussion_r40148686>:
>
> > -		}
> > -	} else {
> > -		safeSetGoError(call.Reply, proto.NewRangeKeyMismatchError(header.Key, header.EndKey, nil))
> > +	key, endKey := keys.Range(ba)
> > +	rng := db.store.LookupReplica(key, endKey)
> > +	if rng == nil {
> > +		return nil, proto.NewError(proto.NewRangeKeyMismatchError(key, endKey, nil))
> > +	}
> > +	ba.RangeID = rng.Desc().RangeID
> > +	replica := rng.GetReplica()
> > +	if replica == nil {
> > +		return nil, proto.NewError(util.Errorf("own replica missing in range"))
> > +	}
> > +	ba.Replica = *replica
> > +	reply, pErr := db.store.ExecuteCmd(ctx, &ba)
> > +	br, _ := reply.(*proto.BatchResponse)
>
> not if it's nil (i.e. on error).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2622/files#r40148686>.
>

tamird commented:
LGTM. I look forward to the death of {,un}wrap*

mrtracy commented:
LGTM

tbg committed:
sauced-up proto.ErrorUnexpectedlySet

and other minor comments by @tamird
